tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/running/HeadlessRunner.java

```diff
@@ -0,0 +1,144 @@
+	 * @param locationToRun
+	 *            location of the code to be executed
+	 * @throws ExitCodeException
+	 *             in cases of errors
+	 */
+	public void startRunner(String runner, String implementationId, String systemLoader, URI locationToRun)
+			throws ExitCodeException {
+
+		IRunnerDescriptor runnerDescriptor = checkRunner(runner);
+		logger.info("Using runner :" + runnerDescriptor.getId());
+
+		RunConfiguration runConfiguration = null;
+		try {
+			runConfiguration = runnerFrontEnd.createConfiguration(runnerDescriptor.getId(), implementationId,
+					systemLoader, locationToRun);
+		} catch (java.lang.IllegalStateException e2) {
+			logger.error(Throwables.getStackTraceAsString(e2));
+			throw new ExitCodeException(EXITCODE_RUNNER_STOPPED_WITH_ERROR,
+					"Cannot create run configuration.", e2);
+		}
+
+		try {
+			Process process = runnerFrontEnd.run(runConfiguration);
+
+			int exit = process.waitFor();
+
+			if (exit != 0) {
+				throw new ExitCodeException(EXITCODE_RUNNER_STOPPED_WITH_ERROR,
+						"The spawned runner '" + runnerDescriptor.getId()
+								+ "' exited with code=" + exit);
+			}
+
+		} catch (InterruptedException e1) {
+			logger.error(Throwables.getStackTraceAsString(e1));
+			throw new ExitCodeException(EXITCODE_RUNNER_STOPPED_WITH_ERROR,
+					"The spawned runner exited by throwing an exception", e1);
+		}
+	}
+
+	/**
+	 * Reads the runner name/ID from the --runWith command line option and checks for validity and, if successful,
+	 * returns the {@link IRunnerDescriptor} for the given name/ID. Otherwise, an {@link ExitCodeException} is thrown.
+	 *
+	 * @return descriptor of runner selected via command line option --runWith
+	 * @throws ExitCodeException
+	 *             in Error cases
+	 */
+	private IRunnerDescriptor checkRunner(String runner) throws ExitCodeException {
+		final List<IRunnerDescriptor> matchingRunnerDescs = findRunnerById(runner);
+
+		if (matchingRunnerDescs.isEmpty()) {
+			throw new ExitCodeException(EXITCODE_RUNNER_NOT_FOUND, "no runner found for id: " + runner);
+		} else if (matchingRunnerDescs.size() > 1) {
+			throw new ExitCodeException(EXITCODE_RUNNER_NOT_FOUND, "several runners match given id \"" + runner
+					+ "\":\n\t" + Joiner.on("\n\t").join(matchingRunnerDescs));
+		}
+
+		return matchingRunnerDescs.get(0);
+	}
+
+	private List<IRunnerDescriptor> findRunnerById(String runnerId) {
+		if (runnerId == null || runnerId.trim().isEmpty())
+			return Collections.emptyList();
+		// 1st attempt: look for exact match
+		final IRunnerDescriptor rd = runnerRegistry.getDescriptors().get(runnerId);
+		if (rd != null)
+			return Collections.singletonList(rd);
+
+		logger.warn("Could not find runner by ID: " + runnerId + ", switching to fuzzy search.");
+
+		// 2nd attempt: look for sloppy match (but full segments required!)
+		final int r_len = runnerId.length();
+		final List<IRunnerDescriptor> matchingRDs = runnerRegistry.getDescriptors().values().stream()
```

Comment:
See comment below: Same here. Also: Can some code be shared/re-used here? It looks similar.